includes/class-sensei-enrolment-calculation-scheduler.php

```diff
@@ -0,0 +1,137 @@
+<?php
+/**
+ * File containing the class Sensei_Enrolment_Calculation_Scheduler.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * The Sensei_Enrolment_Calculation_Scheduler is responsible for running jobs of user enrolment calculations. It is set
+ * up to run once after a Sensei version upgrade or when Sensei_Course_Enrolment_Manager::get_site_salt() is updated.
+ */
+class Sensei_Enrolment_Calculation_Scheduler {
+	const CALCULATION_VERSION_OPTION_NAME = 'sensei-scheduler-calculation-version';
+
+	/**
+	 * Number of seconds between each job run.
+	 *
+	 * @var integer
+	 */
+	private $interval;
+
+	/**
+	 * Number of users for each job run.
+	 *
+	 * @var integer
+	 */
+	private $batch_size;
+
+
+	/**
+	 * Sensei_Enrolment_Calculation_Scheduler constructor.
+	 *
+	 * @param integer $interval The scheduler's interval.
+	 * @param integer $batch_size The scheduler's batch size.
+	 */
+	public function __construct( $interval, $batch_size ) {
+		$this->interval   = $interval;
+		$this->batch_size = $batch_size;
+
+		// phpcs:ignore WordPress.WP.CronInterval.ChangeDetected -- Job size small.
+		add_filter( 'cron_schedules', [ $this, 'add_schedule' ] );
+		add_action( 'sensei_calculate_enrolments', [ $this, 'calculate_enrolments' ], 10, 0 );
+		add_action( 'init', [ $this, 'start' ], 101 );
+	}
+
+	/**
+	 * This method starts the scheduler if the run is not completed for the current Sensei version.
+	 */
+	public function start() {
+		if ( get_option( self::CALCULATION_VERSION_OPTION_NAME ) === Sensei_Course_Enrolment_Manager::get_enrolment_calculation_version() ) {
+			return;
+		}
+
+		if ( ! wp_next_scheduled( 'sensei_calculate_enrolments' ) ) {
```

Comment:
The only thing that concerns me about not making this class a singleton and having the configured internal state for `$interval` and `$batch_state` is this scheduled event name (and the custom interval) is not unique so things will explode a bit if someone were to create a second instance of this class with different config. One solution would be to make those unique with some kind of hash of the internal state, but I don't know if we'd ever want to instantiate with different config.

Reply:
I see your point. When I thought of how the solution should look like, I imagined the scheduler to be something that could possibly be reused to recalculate the enrolment for possibly different user groups/scenarios. In practice this proved a bit complex to implement so I didn't bother until we have the actual need for it.

In order to have it reusable we should have unique hook name, event name, the option name and value for the scheduler progress and the user meta. We could also turn it into a singleton or leave it like this and have the risk of someone creating a new instance. Which option do you think is best?

Reply:
I think singleton just to add a layer of protection against multiple instances. I think for a reusable version we could eventually move some of this to a superclass, but I think each final instance would still be a singleton.

Reply:
Fixed in b2d645738314d88a9abedf0e8e8d0fc4951ea43b